wp-includes/sqlite-ast/class-wp-sqlite-information-schema-reconstructor.php

```diff
@@ -0,0 +1,644 @@
+			)
+		)->fetchAll( PDO::FETCH_COLUMN );
+	}
+
+	/**
+	 * Get the names of all tables recorded in the information schema.
+	 *
+	 * @return string[] The names of tables in the information schema.
+	 */
+	private function get_information_schema_table_names(): array {
+		$tables_table = $this->schema_builder->get_table_name( false, 'tables' );
+		return $this->driver->execute_sqlite_query(
+			"SELECT table_name FROM $tables_table ORDER BY table_name"
+		)->fetchAll( PDO::FETCH_COLUMN );
+	}
+
+	/**
+	 * Get a map of parsed CREATE TABLE statements for WordPress tables.
+	 *
+	 * When reconstructing the information schema data for WordPress tables, we
+	 * can use the "wp_get_db_schema()" function to get accurate CREATE TABLE
+	 * statements. This method parses the result of "wp_get_db_schema()" into
+	 * an array of parsed CREATE TABLE statements indexed by the table names.
+	 *
+	 * @return array<string, WP_Parser_Node> The WordPress CREATE TABLE statements.
+	 */
+	private function get_wp_create_table_statements(): array {
+		if ( ! defined( 'ABSPATH' ) ) {
+			return array();
+		}
+		if ( file_exists( ABSPATH . 'wp-admin/includes/schema.php' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/schema.php';
+		}
+		if ( ! function_exists( 'wp_get_db_schema' ) ) {
+			throw new Exception( 'The "wp_get_db_schema()" function was not defined.' );
+		}
+
+		// Get schema for global tables and the main site.
+		$schema = wp_get_db_schema();
+
+		// For multisite installs, add schema definitions for all sites.
+		if ( is_multisite() ) {
+			$site_ids = get_sites(
+				array(
+					'fields' => 'ids',
+					'number' => PHP_INT_MAX,
+				)
+			);
+			foreach ( $site_ids as $site_id ) {
+				$schema .= wp_get_db_schema( 'blog', $site_id );
+			}
+		}
+
+		// Parse the schema.
+		$parser    = $this->driver->create_parser( $schema );
+		$wp_tables = array();
+		while ( $parser->next_query() ) {
+			$ast = $parser->get_query_ast();
+			if ( null === $ast ) {
+				throw new WP_SQLite_Driver_Exception( $this->driver, 'Failed to parse the MySQL query.' );
+			}
+
+			$create_node = $ast->get_first_descendant_node( 'createStatement' );
+			if ( $create_node && $create_node->has_child_node( 'createTable' ) ) {
+				$name_node = $create_node->get_first_descendant_node( 'tableName' );
+				$name      = $this->unquote_mysql_identifier(
+					substr( $schema, $name_node->get_start(), $name_node->get_length() )
+				);
+
+				$wp_tables[ $name ] = $create_node;
+			}
+		}
+		return $wp_tables;
+	}
+
+	/**
+	 * Generate a MySQL CREATE TABLE statement from an SQLite table definition.
+	 *
+	 * @param  string $table_name The name of the table.
+	 * @return string             The CREATE TABLE statement.
+	 */
+	private function generate_create_table_statement( string $table_name ): string {
+		// Columns.
+		$columns = $this->driver->execute_sqlite_query(
+			sprintf( 'PRAGMA table_xinfo("%s")', $table_name )
```

Comment:
Ditto my point from earlier about safe vs unsafe inputs. Can we either explicitly quote this, even if we don't have to?